sources/file.go

```diff
@@ -0,0 +1,219 @@
+package sources
+
+import (
+	"bufio"
+	"bytes"
+	"context"
+	"fmt"
+	"io"
+	"os"
+	"path/filepath"
+	"strings"
+
+	"github.com/h2non/filetype"
+	"github.com/mholt/archives"
+
+	"github.com/zricethezav/gitleaks/v8/logging"
+)
+
+const defaultBufferSize = 100 * 1_000 // 100kb
+const InnerPathSeparator = "!"
+
+type seekReaderAt interface {
+	io.ReaderAt
+	io.Seeker
+}
+
+// File implements Source for scanning a reader with a path
+type File struct {
+	// Content provides a reader to the file's content
+	Content io.Reader
+	// Path is the resolved real path of the file
+	Path string
+	// Symlink represents a symlink to the file if that's how it was discovered
+	Symlink string
+	// Buffer is used for reading the content in chunks
+	Buffer []byte
+	// outerPaths is the list of container paths (e.g. archives) that lead to
+	// this file
+	outerPaths []string
+}
+
+// Fragments yields fragments for the this source
+func (s *File) Fragments(yield FragmentsFunc) error {
+	ctx := context.Background()
+	format, _, err := archives.Identify(ctx, s.Path, nil)
+
+	if err == nil && format != nil {
+		if extractor, ok := format.(archives.Extractor); ok {
+			return s.extractorFragments(ctx, extractor, s.Content, yield)
+		}
+		if decompressor, ok := format.(archives.Decompressor); ok {
+			return s.decompressorFragments(decompressor, s.Content, yield)
+		}
+		logging.Warn().Str("path", s.Path).Msg("skipping unkown archive type")
+	}
+
+	return s.fileFragments(s.Content, yield)
+}
+
+// extractorFragments recursively crawls archives and yields fragments
+func (s *File) extractorFragments(ctx context.Context, extractor archives.Extractor, reader io.Reader, yield FragmentsFunc) error {
+	if _, isSeekReaderAt := reader.(seekReaderAt); !isSeekReaderAt {
+		switch extractor.(type) {
+		case archives.SevenZip, archives.Zip:
+			tmpfile, err := os.CreateTemp("", "gitleaks-archive-")
```

Comment:
Could we create a single scan directory and nest things underneath?

https://github.com/gitleaks/gitleaks/pull/1841#discussion_r2071557965

Reply:
@rgmz should be doable, but what would be the benefit? Without the parent directory it currently makes cleanup cleaner. 

(Not saying no or pushing back, just trying to get a feel for why ^_^)